### PR TITLE
Stop using the CompilerClassLoader in sbt-bridge

### PR DESCRIPTION
**Pull request by adpi2**
Get rid of the hazardous `CompilerClassLoader`.

As a counterpart it makes the scala compiler bridge incompatible with sbt 1.3.x and Mill older than `0.9.3-21-002361`.

EDIT:
My initial goal was to get rid of the `CompilerClassLoader` but it is probably safer to transition smoothly. So this PR only tries to not use the `CompilerClassLoader`. When the class loader appears to be badly configured, it prints a warning and falls back to the old technique:

```java
    if(isClassLoaderValid()) {
      return new CachedCompilerImpl(options, output);
    } else {
      log.warn(() -> "The compiler class loader is badly configured.");
      // old implementation
    }
```

I tested on:
- sbt `1.3.13` with sbt-dotty `1.4.6`:
```
sbt:foo> compile
[info] Compiling 1 Scala source to /home/piquerez/adpi2/foo/target/scala-3.0.0-RC1/classes ...
[warn] The compiler class loader is badly configured.
[success] Total time: 3 s, completed Jan 6, 2021, 11:54:29 AM
```
- Mill `0.9.3`
```
./millw foo.compile
[27/27] foo.compile 
[info] Compiling 1 Scala source to /home/piquerez/adpi2/foo/out/foo/compile/dest/classes ...
[warn] The compiler class loader is badly configured.
[info] Done compiling.
```
- Bloop `1.4.6`
```
bloop compile foo
Compiling root (1 Scala source)
[W] The compiler class loader is badly configured.
Compiled root (1961ms)
```
The warning goes away with sbt `1.4.x` (sbt-dotty `1.5.1`) and Mill `0.9.4`.

[test_sbt]

### CODE REVIEW

**Comment by adpi2** [comment]
@eed3si9n, @lefou, @lihaoyi 

Just so you know that this PR will make the compiler bridge of the next Scala 3 version incompatible with sbt `1.3.x` and Mill prior to `0.9.3-21-002361`.

**Comment by adpi2** [comment]
Unfortunately it is also incompatible with latest Bloop `v1.4.6-15-209c2a5c`, so I am going to adapt the class loader in Bloop as well.

**Comment by eed3si9n** [comment]
acknowledged.

Is there a try-catch or reflection hack you can implement to use the old technique as a fallback?

**Comment by adpi2** [comment]
@eed3si9n 
Yes, that's probably a good thing to do.

File: sbt-bridge/src/xsbt/CompilerInterface.java [hunk 1]
```diff
@@ -6,36 +6,44 @@
...
-    // The classloader that sbt uses to load the compiler bridge is broken
-    // (see CompilerClassLoader#fixBridgeLoader for details). To workaround
-    // this we construct our own ClassLoader and then run the following code
-    // with it:
-    //   new CachedCompilerImpl(options, output)
+  public CachedCompiler newCompiler(String[] options, Output output, Logger initialLog, Reporter initialDelegate) throws java.lang.Exception {
+    if(isClassLoaderValid()) {
+      return new CachedCompilerImpl(options, output);
+    } else {
+      initialLog.warn(() -> "The compiler class loader is badly configured.");
```
Comment (smarter) [inline sbt-bridge/src/xsbt/CompilerInterface.java:25]: This message should mention the minimum version users should upgrade to for sbt/sbt-dotty/mill/bloop.

**Review by smarter — APPROVED** [review]

File: sbt-bridge/src/xsbt/CompilerInterface.java [hunk 2]
```diff
@@ -6,36 +6,44 @@
...
+        ClassLoader fixedLoader = CompilerClassLoader.fixBridgeLoader(bridgeLoader);
+        Class<?> cciClass = fixedLoader.loadClass("xsbt.CachedCompilerImpl");
+        return (CachedCompiler) cciClass.getConstructors()[0].newInstance(options, output);
+      } catch (ClassNotFoundException | InstantiationException | IllegalAccessException | InvocationTargetException e) {
+        throw new RuntimeException(e);
+      }
+    }
+  }
 
+  private Boolean isClassLoaderValid() {
```
Comment (allanrenucci) [inline sbt-bridge/src/xsbt/CompilerInterface.java:40]:
```suggestion
  private boolean isClassLoaderValid() {
```

**Pull request merged**